source/core_mqtt_serializer.c

```diff
@@ -2623,6 +2628,32 @@ MQTTStatus_t MQTT_GetIncomingPacketTypeAndLength( TransportRecv_t readFunc,
 
 /*-----------------------------------------------------------*/
 
+MQTTStatus_t MQTT_UpdateDuplicatePublishFlag( uint8_t * pHeader , bool set)
+{
+    MQTTStatus_t status = MQTTSuccess;
+
+    if(pHeader == NULL)
+    {
+        status = MQTTBadParameter;
+    } 
+    else if( ( ( *pHeader ) & MQTT_PACKET_TYPE_PUBLISH ) == 0 )
+    {
+        status = MQTTBadParameter;
+    } 
+    else if( set )
```

Comment:
Just wanted to know if this is a MISRA violation specific change or in general `set == true` if preffered more? because according to me, if the variable within the if condition is of type boolean then comparing it with `true` or `false` is less efficient as compared to using it directly.

Reply:
Both will compile down to the same thing. Misra allows either as long as set is 'essentially boolean' as is defined in one of the appendixes. For pointers, `== NULL` is required as the if condition expression must be essentially boolean but a pointer isnt. The result of `==` is though.

Reply:
Agree with Archit.

`if(set == true) and `if(set)` both lead to the compiler generating same code because the latter is intrinsic check against being true or false and the previous is an extrinsic check. They make our life easier - not the compilers :) 

But, for  the sake of consistency, we should make it as I suggested.

Reply:
Well `if(x)` semantically jumps over the section if x is zero. but compiler will see that `== true` is not changing the result and will drop it.

Reply:
for the iovec thing, I think it makes sense as allows more flexibility in storage. I'd probably use contiguous memory, but can do that by just passing a single elem iovec on return and concatenating the elems on receiving.

Reply:
`if( a )` compiles to this assembly:
```asm
        ldrb    w0, [sp, 31]
        and     w0, w0, 255
        and     w0, w0, 1
        cmp     w0, 0        # This is the line to focus on.
        beq     .L2
```

and `if(a == true)` compiles to this:
```asm
        ldrb    w0, [sp, 31]
        and     w0, w0, 255
        and     w0, w0, 1
        cmp     w0, 0        # This is the line to focus on.
        beq     .L3
```

The only difference in both is the branch location - as it were two different branches.